Allow PackageSourceProvider to avoid subscribing to configuration change event

zivkan opened the pull request:
## Bug

Fixes: NuGet/Home#8471
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Overload constructor to allow opting out of subscribing to the events. Note: this is a short/medium term fix, NuGet/Home#8479 is needed to fix it properly.

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

zivkan committed:
Allow PackageSourceProvider to avoid subscribing to configuration change event

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs:
```diff
@@ -108,6 +108,7 @@ private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
 
             if (hasChanged)
             {
+                var invokers = SettingsChanged.GetInvocationList();
```
What was your plan with this unused local?

zivkan replied:
oops, that was just for debugging, to make sure when I closed VS that my other changes worked as expected.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs:
```diff
@@ -14,7 +14,7 @@ namespace NuGet.PackageManagement.VisualStudio
     [Export(typeof(ISettings))]
     [PartCreationPolicy(CreationPolicy.Shared)]
-    public class VSSettings : ISettings
+    public class VSSettings : ISettings, IDisposable
```
Read [this](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8#idisposable-and-the-inheritance-hierarchy).

Ideally, `VSSettings` would be `sealed`, but if it can't be, provide a `protected virtual void Dispose(bool)` method and make sure to call it in `Dispose()`.

zivkan replied:
serves me right for copying how some of our other classes implement IDisposable. I'll do it properly and create a task to do the others later.

dtivel commented:
Rebase on latest dev to pick up the [fix](https://github.com/NuGet/NuGet.Client/commit/2f9c60e1708347789766c67cd981987784d9edef) for your test failures.

zivkan commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -18,17 +18,34 @@ public class PackageSourceProvider : IPackageSourceProvider
+        {
+        }
+
+        public PackageSourceProvider(
```
Does anyone have any ideas how we can discourage customers from using this overload, given NuGet/Home#8479 should remove the need for it?

My best idea is to mark it as obsolete, but then we'll get warnings where we use it unless we disable the obsolete warning for those lines, which doesn't feel better than not doing anything.

dtivel replied:
Warn in XML doc comments?

nkolev92 replied:
mark as obsolete, say it's going away, suppress on our end, fix in 6.0? 
That's the best I have. 

Also, you can't not disable the warning, we have WarningsAsErrors. :)

zivkan committed:
work

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSettingsTask.cs:
```diff
@@ -133,7 +133,7 @@ public override bool Execute()
                     () => MSBuildRestoreUtility.ContainsClearKeyword(RestoreSources) ? Array.Empty<string>() : null,
                     () => RestoreSources?.Select(MSBuildRestoreUtility.FixSourcePath).Select(e => UriUtility.GetAbsolutePathFromFile(ProjectUniqueName, e)).ToArray(),
-                    () => (new PackageSourceProvider(settings)).LoadPackageSources().Where(e => e.IsEnabled).Select(e => e.Source).ToArray());
+                    () => (new PackageSourceProvider(settings, enablePackageSourcesChangedEvent: false)).LoadPackageSources().Where(e => e.IsEnabled).Select(e => e.Source).ToArray());
```
In many places in this PR there are instantiations of `PackageSourceProvider` (with the new argument) just to call `LoadPackageSources()`.

Consider having a static (internal?) method (`PackageSourceProvider.LoadPackageSources(ISettings)`) which never subscribes events and eliminates the allocation for the throwaway `PackageSourceProvider` instance.  Within `PackageSourceProvider` you can share implementation with the constructor or leave it to https://github.com/NuGet/Home/issues/8479 fix.

zkat replied:
There's something that makes me a bit uncomfortable about the default being the more resource-intensive one and having to pepper this extra parameter all throughout the code (and remember to do so), esp. since there doesn't seem any disposer that will remove the event right now.

I think this would very well be fixed by @dtivel's suggestion above, too. Are all the relevant cases where the change event is disabled the ones where `#LoadPackageSources()` is called?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/ExtensibleSourceRepositoryProvider.cs:
```diff
@@ -59,10 +59,10 @@ private void EnsureInitialized()
                     _initialized = true;
 
-                    _packageSourceProvider = new Configuration.PackageSourceProvider(_settings.Value);
+                    _packageSourceProvider = new Configuration.PackageSourceProvider(_settings.Value, true);
```
Use a named parameter for `true`.

zivkan committed:
minor improvement because tests failed, so I have to rerun tests again and hope it's less flakey.

zkat commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -18,17 +18,34 @@ public class PackageSourceProvider : IPackageSourceProvider
-            Settings.SettingsChanged += (_, __) => { OnPackageSourcesChanged(); };
+            if (enablePackageSourcesChangedEvent)
+            {
+                Settings.SettingsChanged += (_, __) => { OnPackageSourcesChanged(); };
```
When is this disposed of?

zivkan replied:
it's not. NuGet/Home#8479 needs to fix it. We can't fix it without breaking backwards compat for customers using our package.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/ExtensibleSourceRepositoryProvider.cs:
```diff
@@ -16,7 +16,7 @@ namespace NuGet.PackageManagement.VisualStudio
+    public sealed class ExtensibleSourceRepositoryProvider : ISourceRepositoryProvider, IDisposable
     {
 
         // TODO: add support for reloading sources when changes occur
```
This TODO is an eye sore lol.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/ExtensibleSourceRepositoryProvider.cs:
```diff
@@ -59,10 +59,10 @@ private void EnsureInitialized()
                     _initialized = true;
 
-                    _packageSourceProvider = new Configuration.PackageSourceProvider(_settings.Value);
+                    _packageSourceProvider = new Configuration.PackageSourceProvider(_settings.Value, enablePackageSourcesChangedEvent: true);
```
using NuGet.Configuration;

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PushCommand.cs:
```diff
@@ -98,7 +98,7 @@ public static void Register(CommandLineApplication app, Func<ILogger> getLogger)
                     }
 
-                    var sourceProvider = new PackageSourceProvider(XPlatUtility.CreateDefaultSettings());
+                    var sourceProvider = new PackageSourceProvider(XPlatUtility.CreateDefaultSettings(), enablePackageSourcesChangedEvent: true);
```
This one seems weird. 
Do these events fire in commandline scenarios?

zivkan replied:
no, it was a mistake.

zkat commented on src/NuGet.Clients/NuGet.CommandLine/PackageSourceBuilder.cs:
```diff
@@ -1,13 +1,15 @@
+        internal static PackageSourceProvider CreateSourceProvider(ISettings settings)
         {
-            return new Configuration.PackageSourceProvider(settings);
+            return new PackageSourceProvider(settings, enablePackageSourcesChangedEvent: false);
```
Is this because all the consumers are CLI commands?

zivkan replied:
No, it was a mistake. But ultimately it doesn't matter because the CLI commands are short lived processes, so leaks are limited in impact. I largely made the change so I could use IntelliCode to find all references to the constructor without the bool parameter, to ensure I find all usage that does matter (which happened to only be 1). Anyway, fixed now.

zkat commented on src/NuGet.Core/NuGet.Protocol/SourceRepositoryProvider.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Copyright (c) .NET Foundation. All rights reserved.
```
This looks like a BOM removal. Or addition. Not sure.

zivkan replied:
Indeed. GitHub doesn't show it, but `git diff` on the command line shows it clearly. Our `.editorconfig` defines `charset = utf-8`, which is the no BOM charset.

nkolev92 reviewed (commented):
Looks great. 
I like Damon's idea about a utiility method. 

Although I think it can't be internal, so only do it if there's a property utility class.

nkolev92 reviewed (approved):
Meant to approve.

zivkan committed:
more work

zivkan committed:
Make static PackageSourceProvider.LoadPackageSources

zivkan committed:
Mark new PackageSourceProvider constructor as obsolete

It's a short-medium term fix. Done this way to avoid breaking API/ABI in a minor release.

dtivel commented on src/NuGet.Core/NuGet.Protocol/SourceRepositoryProvider.cs:
```diff
@@ -18,7 +18,9 @@ public class SourceRepositoryProvider : ISourceRepositoryProvider
         public SourceRepositoryProvider(ISettings settings, IEnumerable<Lazy<INuGetResourceProvider>> resourceProviders)
-            : this(new PackageSourceProvider(settings), resourceProviders)
+#pragma warning disable CS0618 // Type or member is obsolete
+            : this(new PackageSourceProvider(settings, enablePackageSourcesChangedEvent: false), resourceProviders)
```
`false` looks like a bug.  See the new line 40 below.

    packageSourceProvider.PackageSourcesChanged += (sender, e) => { Init(); };

zivkan committed:
work

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -18,21 +18,44 @@ public class PackageSourceProvider : IPackageSourceProvider
+        {
+        }
+
+        public PackageSourceProvider(
```
What is the long term strategy for the PackageSourceProvider?

Will it subscribe or not subscribe to events? 

Are the obsoletes maybe flipped or something?

zivkan replied:
> Will it subscribe or not subscribe to events?

As a team, we need to decide. Basically, do we want to support detecting config file changes and auto-reload? Maybe we can do it just in the VS client code, and the packages don't need it. But until we decide, I don't know whether the events themselves are obsolete or not.

nkolev92 replied:
>  Basically, do we want to support detecting config file changes and auto-reload? 

Eventually, but I don't necessarily think this would prevent us from doing that. 

It was just difficult for me to understand what codepath exactly we are trying to deprecate. 
Maybe you can elaborate on that?

zivkan replied:
I want to deprecate the constructors that have a `enablePackageSourcesChangedEvent` parameter, or anything that calls the deprecated constructor.

In the long term, either the class always subscribes to the event, and the class is IDisposable. Or the class never subscribes to the event. In either case, the constructors with `enablePackageSourcesChangedEvent` will not be needed.

nkolev92 replied:
Makes sense.

nkolev92 reviewed (approved):
Approval still stands from earlier :)
